### PR TITLE
add flag for skipping files inside vendor directory

### PR DESCRIPTION
**Pull request by artuross**
Added flag `--skip-vendor` as suggested in #129 for skipping files inside `vendor` directory.

### CODE REVIEW

File: pkg/gci/gci.go [hunk 1]
```diff
@@ -110,6 +115,12 @@ func processingFunc(file io.FileObj, cfg config.Config, formattingFunc fileForma
 	}
 }
 
+// A more sophisticated version would check for the location of go.mod file,
+// but it is probably not needed.
+func isInsideVendorDir(path string) bool {
+	return strings.HasPrefix(path, "vendor/") || strings.Contains(path, "/vendor/")
```
[the remark this answers is not in the document]
Comment (daixiang0) [reply pkg/gci/gci.go:121]: @artuross cloud you take a look and update?

[the remark this answers is not in the document]
Comment (artuross) [reply pkg/gci/gci.go]: Sorry for the delay on this. I've improved the vendor check.

File: pkg/gci/gci.go [hunk 2]
```diff
@@ -90,6 +91,10 @@ func ProcessFiles(fileGenerator io.FileGeneratorFunc, cfg config.Config, fileFun
 		return err
 	}
 	for _, file := range files {
+		if cfg.SkipVendor && isInsideVendorDir(file.Path()) {
```
Comment (daixiang0) [inline pkg/gci/gci.go:94]: Could you modify `GoFilesInPathsGenerator` to skip vendor directly? If so, we will reduce check times.

File: pkg/io/search.go [hunk 3]
```diff
@@ -42,6 +42,36 @@ func findFilesForDirectory(dirPath string, fileCheckFun fileCheckFunction) ([]st
...
+
+		if prevPath == path {
+			break
+		}
+	}
+
+	return true
+}
+
+func all(checkFuncs ...fileCheckFunction) fileCheckFunction {
```
Comment (daixiang0) [inline pkg/io/search.go:67]: Please name as `checkChains` or something more meaningful.

Comment (artuross) [reply pkg/io/search.go]: Changed.

File: pkg/io/file.go [hunk 4]
```diff
@@ -39,8 +39,13 @@ func (a FileGeneratorFunc) Combine(b FileGeneratorFunc) FileGeneratorFunc {
 	}
 }
 
-func GoFilesInPathsGenerator(paths []string) FileGeneratorFunc {
-	return FilesInPathsGenerator(paths, isGoFile)
+func GoFilesInPathsGenerator(paths []string, skipVendor bool) FileGeneratorFunc {
+	checkFunc := isGoFile
+	if skipVendor {
+		checkFunc = all(isGoFile, isOutsideVendorDir)
```
Comment (daixiang0) [inline pkg/io/file.go:45]: Still `all()`?

Comment (daixiang0) [reply pkg/io/file.go:45]: ping @artuross

**Comment by emilien-puget** [comment]
Hi, this feature is really a blocker for me to use this library, how could we move forward on this topic ?

File: pkg/io/file.go (hunk 4 above)
Comment (artuross) [reply pkg/io/file.go]: @daixiang0 done

**Review by daixiang0 — APPROVED** [review]
LGTM

**Comment by emilien-puget** [comment]
thank you all for your swift response, what are the next steps to integrate this new feature to golangci-lint, if any ?

**Comment by artuross** [comment]
@emilien-puget need to wait for the next release of `gci`, then update the version in `golangci-lint` and add the flag in the [settings](https://github.com/golangci/golangci-lint/blob/master/pkg/config/linters_settings.go#L388) and later pass the flag to `io.GoFilesInPathsGenerator`.

**Pull request merged**